torch/nested/_internal/nested_tensor.py

```diff
@@ -132,38 +129,75 @@ def __new__(
 
         return r
 
-    def __init__(self, values, offsets, *, lengths=None, **kwargs):
+    def __init__(self, values, metadata, non_contig_offsets=None, **kwargs):
         super().__init__()
 
         self._values = values
-        self._offsets = offsets
-        self._lengths = lengths
-
-        # holds properties that are computed lazily
-        self._metadata_cache = kwargs.get("_metadata_cache") or {}
+        self._non_contig_offsets = non_contig_offsets
 
         # collapsed ragged dim must always be dynamic
         torch._dynamo.maybe_mark_dynamic(self, self._ragged_idx)
         torch._dynamo.maybe_mark_dynamic(self._values, self._ragged_idx - 1)
 
-        # min / max sequence length should be dynamic if present
-        max_seqlen_tensor = self._metadata_cache.get("max_seqlen", None)
-        if max_seqlen_tensor is not None:
-            torch._dynamo.mark_dynamic(max_seqlen_tensor, 0)
-        min_seqlen_tensor = self._metadata_cache.get("min_seqlen", None)
-        if min_seqlen_tensor is not None:
-            torch._dynamo.mark_dynamic(min_seqlen_tensor, 0)
-
-    def values(self):
+    def values(self) -> torch.Tensor:
         # dispatch to get proper view relationship
         return torch._nested_get_values(self)  # type: ignore[attr-defined]
 
-    def offsets(self):
-        return self._offsets
+    # Public APIs for lengths/offsets returns plain tensor for BC reasons, e.g. user
+    # can access .data_ptr() on it, etc.
+    # .lengths() and .offsets() do not automatically do conversion between lengths
+    # and offsets. If the requested variant does not exist, then None is returned.
+    def offsets(self) -> torch.Tensor:
+        ret = self._offsets
+        assert ret is not None
+        return ret
 
-    def lengths(self):
+    def lengths(self) -> Optional[torch.Tensor]:
         return self._lengths
 
+    def __getattr__(self, name) -> Optional[torch.Tensor]:
+        if name not in SOURCE_FIELDS + EXTRA_FIELDS + ("_lengths", "_offsets"):
+            raise AttributeError(
+                f"{type(self).__name__} object has no attribute '{name}'"
+            )
+
+        if name == "_offsets":
+            if self._non_contig_offsets is not None:
+                # non-contig case
+                return self._non_contig_offsets
+            # contig case
+            if self.is_cpu:
+                return self._host_offsets
+            else:
+                return self._device_offsets
+
+        elif name == "_lengths":
+            if self.is_cpu:
+                return self._host_lengths
+            else:
+                return self._device_lengths
+
+        return getattr(self._metadata, name)
```

Comment:
gross, I'm not crazy about adding a custom getattr. Isn't it possible to `_offsets` / `_lengths` via `@properties`?

also there aren't -that- many extra fields to define properties for, right?

Reply:
Happy to use properties instead. There would be close to 10, but maybe that's not a big deal.

Reply:
One thing I like about getattr though is that bc it is more programatic, it may be simpler for the user to define their own cache entries. But maybe not important until we actually try to go and support that.

Reply:
> One thing I like about getattr though is that bc it is more programatic

I don't disagree with this, I'm just (possibly irrationally) scared of using a heavy tool like a custom getattr and how that might interact with other subsystems (e.g. Dynamo).

Reply:
Okay I'm less scared since it's `__getattr__` and not `__getattribute__`. I'll assume the tests coverage address Dynamo compatibility

Reply:
> My opinion is that properties should be used where we're not taking advantage of the programmatic nature. Specifically, for _offsets and _lengths. The other stuff can be handled programmatically as done now.

This is a good point and should clean up this function a bit.